types/field_type.go

```diff
@@ -185,7 +185,7 @@ func DefaultParamTypeForValue(value interface{}, tp *FieldType) {
 func hasVariantFieldLength(tp *FieldType) bool {
 	switch tp.Tp {
 	case mysql.TypeLonglong, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeBlob,
-		mysql.TypeBit, mysql.TypeDuration, mysql.TypeNewDecimal, mysql.TypeEnum, mysql.TypeSet:
+		mysql.TypeBit, mysql.TypeDuration, mysql.TypeEnum, mysql.TypeSet:
```

Comment:
Will this incur same problem for Decimal type as https://github.com/pingcap/tidb/issues/8644?

Reply:
Thanks, I will add a test for it.

Reply:
PTAL @eurekaka

Reply:
What does your test mean in the PR description? have you enabled the prepare-plan-cache for the test? your test can only check the flen constraint of the column definition?

The problem of issue/8644 is that, during the first execution of the query, the flen is set to 1, which would fail the subsequent executions of the query if their arguments have larger field length than 1. All those field lengths may be in the range of the column type, so those queries should not fail.

Reply:
Also, please add the unit test into TiDB repo.

Reply:
Thanks for your reply, I misunderstood the issue/8644.

> Have you enabled the prepare-plan-cache for the test?

Yes, my test environment opens plan-cache defaulted.

> What does your test mean in the PR description?

Sorry, this is not `plan cache`'s issue, I closed plan-cache but also get the error `Wrong precision`. The description has been updated.

> Will this incur the same problem for Decimal type as #8644?

In the current version, when we build a Decimal column without `M` and `D`, it will be auto-filled by `Decimal(10,0)`. So it can't cause a problem like issue/8644.

> Also, please add the unit test into TiDB repo.

I think `Wrong precision` can't be tested out in TiDB repo, because it is reported by TiFlash.